### PR TITLE
Notifications: use server side notification groups

### PR DESCRIPTION
**Pull request by cjb**
@keybase/react-hackers

Before this PR, the store looked like:
```
settings: {
  notifications: {
    follow: boolean,
    unsubscribedFromAll: boolean,
  }
}
```
After this PR, it looks like:
```
settings: {
  notifications: {
    groups: {
      app_push: {
        settings: {
          follow: boolean,
        },
        unsubscribedFromAll: boolean,
      },
      email: {
        settings: {
          follow: boolean,
        },
        unsubscribedFromAll: boolean,
      },
      ...
    }
  }
}
```
The "groups" (app_push, email, sms) and settings are sent by the server-side.

The display component chooses to display email, and app_push if it exists (which is the case only if you have a phone device).  It would be possible to make the display component iterate through groups without prior knowledge of which ones exist, but that seemed a step too far to me -- and if we wanted to do that we'd need the API server to start sending us more labels, e.g. app_push => "Push notifications" and unsubscribe text.  So this PR data-drives the store, but keeps the display components explicit about what they want to display.

### CODE REVIEW

**Comment by keybase-ci-visdiff** [comment]
The commits ddee2b1f02b8...844f88639312 introduce visual changes on linux. <details><summary>:mag_right: 2 changed</summary>

 * Changed: **Notifications-Unsub_All** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495645427589-ddee2b1f02b8-844f88639312-linux/Notifications-Unsub_All-ddee2b1f02b8.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495645427589-ddee2b1f02b8-844f88639312-linux/Notifications-Unsub_All-844f88639312.png)  
   Notifications-Unsub_All comparison
 * Changed: **Notifications-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495645427589-ddee2b1f02b8-844f88639312-linux/Notifications-Normal-ddee2b1f02b8.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495645427589-ddee2b1f02b8-844f88639312-linux/Notifications-Normal-844f88639312.png)  
   Notifications-Normal comparison
</summary>

File: shared/settings/notifications/index.native.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+import type {Props} from './index'
+
+const MobileNotifications = (props: Props) => (
+  <NativeScrollView style={{...globalStyles.flexBoxColumn, flex: 1}}>
+    <Notifications {...props} />
+  </NativeScrollView>
+)
+
+export default HeaderHoc(MobileNotifications)
```
Comment (chrisnojima) [inline shared/settings/notifications/index.native.js]: no scroll concerns on desktop?

File: shared/settings/notifications/index.js [hunk 2]
```diff
@@ -1,41 +1,62 @@
...
+          onCheck={() => props.onToggleUnsubscribeAll('email')}
           disabled={!props.allowEdit}
-          checked={!!props.unsubscribedFromAll}
+          checked={props.groups.email && !!props.groups.email.unsubscribedFromAll}
           label="Unsubscribe me from all mail"
         />
+
+        {!!props.groups.app_push &&
+          props.groups.app_push.settings &&
+          <Box style={globalStyles.flexBoxColumn}>
```
Comment (chrisnojima) [inline shared/settings/notifications/index.js]: we should make a helper for this or something as its really copy pasta of the above right?

File: shared/settings/notifications/index.js [hunk 3]
```diff
@@ -1,41 +1,65 @@
...
-        <Text type="BodyBig" style={{marginTop: globalMargins.medium}}>Or:</Text>
-        <Checkbox
-          style={{marginTop: globalMargins.small, marginBottom: globalMargins.medium}}
-          onCheck={() => props.onToggleUnsubscribeAll()}
-          disabled={!props.allowEdit}
-          checked={!!props.unsubscribedFromAll}
-          label="Unsubscribe me from all mail"
-        />
+    : <Box style={{...globalStyles.scrollable, padding: globalMargins.small, flex: 1}}>
+        {props.groups.email && props.groups.email.settings && renderGroup('email', props)}
```
Comment (chrisnojima) [inline shared/settings/notifications/index.js]: you might as well make this a stateless component like
```
{props.groups.email && props.groups.email.settings && <Group {...props} />}
```

**Comment by cjb** [comment]
@chrisnojima Thanks, added scrollable for desktop, and a helper for the copypasta.

File: shared/settings/notifications/index.js [hunk 4]
```diff
@@ -1,41 +1,65 @@
...
+    key={s.name}
+    disabled={!props.allowEdit}
+    onCheck={() => props.onToggle(group, s.name)}
+    checked={s.subscribed}
+    label={s.description}
+  />
+)
+
+const renderGroup = (groupName: string, props: Props) => {
+  const labels = {
```
Comment (chrisnojima) [inline shared/settings/notifications/index.js]: I think this stuff should be passed in, and we dont' need to pass props itself in, just what this thing needs. I'd imagine the props this component gets are like
```
<Group
title='Email me:',
unsub='mail' // (so we put the 'Unsubscribe me from ${unsub}')
settings={this.props.groups.app_push.settings}
allowEdit={this.props.allowEdit}
onToggle={settings => this.props. onToggleUnsubscribe(settings) ... etc}
/>

```

**Comment by keybase-ci-visdiff** [comment]
The commits ddee2b1f02b8...e0ab0f48d35a introduce visual changes on linux. <details><summary>:mag_right: 2 changed</summary>

 * Changed: **Notifications-Unsub_All** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495655720878-ddee2b1f02b8-e0ab0f48d35a-linux/Notifications-Unsub_All-ddee2b1f02b8.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495655720878-ddee2b1f02b8-e0ab0f48d35a-linux/Notifications-Unsub_All-e0ab0f48d35a.png)  
   Notifications-Unsub_All comparison
 * Changed: **Notifications-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495655720878-ddee2b1f02b8-e0ab0f48d35a-linux/Notifications-Normal-ddee2b1f02b8.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495655720878-ddee2b1f02b8-e0ab0f48d35a-linux/Notifications-Normal-e0ab0f48d35a.png)  
   Notifications-Normal comparison
</summary>

**Comment by cjb** [comment]
@chrisnojima Makes sense, how does this look?

**Comment by keybase-ci-visdiff** [comment]
The commits ddee2b1f02b8...301dc8b3e58d introduce visual changes on linux. <details><summary>:mag_right: 2 changed</summary>

 * Changed: **Notifications-Unsub_All** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495660350945-ddee2b1f02b8-301dc8b3e58d-linux/Notifications-Unsub_All-ddee2b1f02b8.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495660350945-ddee2b1f02b8-301dc8b3e58d-linux/Notifications-Unsub_All-301dc8b3e58d.png)  
   Notifications-Unsub_All comparison
 * Changed: **Notifications-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495660350945-ddee2b1f02b8-301dc8b3e58d-linux/Notifications-Normal-ddee2b1f02b8.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495660350945-ddee2b1f02b8-301dc8b3e58d-linux/Notifications-Normal-301dc8b3e58d.png)  
   Notifications-Normal comparison
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits ddee2b1f02b8...2f070d8c434c introduce visual changes on linux. <details><summary>:mag_right: 2 changed</summary>

 * Changed: **Notifications-Unsub_All** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495660829640-ddee2b1f02b8-2f070d8c434c-linux/Notifications-Unsub_All-ddee2b1f02b8.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495660829640-ddee2b1f02b8-2f070d8c434c-linux/Notifications-Unsub_All-2f070d8c434c.png)  
   Notifications-Unsub_All comparison
 * Changed: **Notifications-Normal** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495660829640-ddee2b1f02b8-2f070d8c434c-linux/Notifications-Normal-ddee2b1f02b8.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1495660829640-ddee2b1f02b8-2f070d8c434c-linux/Notifications-Normal-2f070d8c434c.png)  
   Notifications-Normal comparison
</summary>

**Comment by cjb** [comment]
Hm!  Removing the flow definitions doesn't seem to work to catch missing props here -- removing the "subscribed={s.subscribed}" line correctly gives a flow warning at the moment:
```
settings/notifications/index.js:42
 42:           <SubscriptionCheckbox
               ^ React element `SubscriptionCheckbox`
  9: const SubscriptionCheckbox = (props: {
                                          ^ property `subscribed`. Property not found in
 42:           <SubscriptionCheckbox
               ^ props of React element `SubscriptionCheckbox`
```
but no longer gives the warning if I remove the types on SubscriptionCheckbox.

**Pull request merged**